test/density_interface.jl

```diff
@@ -0,0 +1,16 @@
+@testset "DensityInterface" begin
+    using DensityInterface
+
+    d_uv_continous = Normal(-1.5, 2.3)
+    d_uv_discrete = Poisson(4.7)
+    d_mv = MvNormal([2.3 0.4; 0.4 1.2])
+
+    for d in (d_uv_continous, d_uv_discrete, d_mv)
+        x = rand(d)
```

Comment:
Maybe add a test with multiple samples as well?

Reply:
You mean multiple samples in one call? I don't think DensityInterface should support this, really. It predates Julia having broadcasting in the design of Distributions - but nowadays, I don't think it's a good idea. And it's not cleanly done in Distributions currently anyway: Multiple samples get passed as a matrix to a multivariate dist, but as a vector of matrices to a matrix-variate dist ...

Reply:
Ok, I've blocked this now, and added some tests for it. Ok with you?

Reply:
Note: I did it this way (blocking instead of restricting to specific array dimensionality in the primary definition of `logdensityof`) to support Distributions that have non-array-like variates (e.g. `ValueShapes.NamedTupleDist`).

Reply:
> You mean multiple samples in one call? I don't think DensityInterface should support this, really. It predates Julia having broadcasting in the design of Distributions - but nowadays, I don't think it's a good idea.

I think we talk about different things here. I don't want to support broadcasting, ie. `logdensityof.((dist,), samples)`. Multiple samples here means `sum(logpdf.((dist,), samples))`, i.e., it just returns the log-likelihood of a data set of `samples`. I think it would be useful to support this with `logdensityof` as it would simplify working with log-likelihoods. It would also be much more difficult to use `logdensityof` in Turing if it is not supported as it is based on calls such as `loglikelihood(Normal(), data)` etc.

> And it's not cleanly done in Distributions currently anyway: Multiple samples get passed as a matrix to a multivariate dist, but as a vector of matrices to a matrix-variate dist ...

There is some inconsistency but only because multivariate data/samples as matrices is so popular. `loglikelihood` also supports arrays (not only vectors) of vectors, so one does not have to use matrices in the evaluation. Additionally, all these implementation details and optimizations of `loglikelihood` are handled by Distributions and hence nothing that DensityInterface has to be concerned about.

Also due to these inconsistencies, it would be very annoying to replace calls of `loglikelihood(dist, samples)` in Turing with something like `sum(logdensityof.((dist,), samples))` - it would just not work for all combinations of distributions and samples, and hence we would have to write our own custom function that dispatches on the types of distributions. And we would have to implement optimizations for specific distributions on our own. It seems much more sane to just rely on Distributions implementing `loglikelihood` in a reasonable and optimized way.

Reply:
> I think we talk about different things here. I don't want to support broadcasting, ie. `logdensityof.((dist,), samples)`. Multiple samples here means `sum(logpdf.((dist,), samples))`, i.e., it just returns the log-likelihood of a data set of `samples`.

Say `dist` is a distribution over some complex type, so "is it an Array or not" isn't enough. This could make it very hard to tell whether what's being passed is intended as a set of samples, vs one sample from a distribution over arrays.

Reply:
I'm not sure what you mean and it's not clear to me how it is related to the discussion. `sum(logpdf.((dist,), samples))` was just a simple illustration for univariate samples to make it clear what I meant with "multiple samples" and that it is not just broadcasting. This is not how `loglikelihood` is implemented for general distributions. And in the implementation of `loglikelihood` for a specific distribution there is no confusion about whether a single sample or a set of samples is passed. However, all these discussions about the implementation of `loglikelihood` are completely irrelevant for the definition of `logdensityof`.

Reply:
Yes, from a dispatch viewpoint this can be awkward, supporting both a single and multiple samples in the same function call. I'm also not sure that it matches the semantics we defined for `logdensityof` well: We're just using `loglikelihood` instead of `logpdf` here because of implementation reasons. But the density we speak of in regard to `logdensityof` is the PDF, and `logdensityof` is supposed to return that density at a single point.

It's of course debatable what "a single point" is: it can of course be a concrete set of samples drawn from the set of all possible sets of samples. Still, I don't think we should say that the likelihood is the natural density of a distribution.

I do however see you points regarding the practical aspects, @devmotion. How about we do this, in addition to defining `logdensityof` for distributions:

```julia
struct Likelihood{D<:Distribution}
    d::D
end

@inline DensityInterface.hasdensity(l::Likelihood) = true

@inline DensityInterface.logdensityof(l::Likelihood, x) = likelihood(l.d, x)
```

I would argue that the PDF and the likelihood are very different things, semantically and mathematically, so we should not use `logdensityof(d, x)` to get the one or the other depending on the type of `x`.

Reply:
Ah, no, that's also not so nice, since the argument of a likelihood density are the parameters of the distribution, not the data.

What does Turing need here, exactly?

Reply:
We need and use `loglikelihood`. There's not much use for a `logdensityof` that only gives us the `logpdf`, due to all the inconsistencies and dispatches that we would have to solve on our own - even though `loglikelihood` already solves them.

Reply:
From
> I think it would be useful to support this with `logdensityof` as it would simplify working with log-likelihoods.

it seemed to me you were suggesting `logdensity` adopt this behavior.

> However, all these discussions about the implementation of `loglikelihood` are completely irrelevant for the definition of `logdensityof`.

I was commenting on something you brought up. Maybe I misunderstood something. I don't assume you have deep knowledge of how MeasureTheory works, and I'm happy to help with sincere misunderstandings. I'd appreciate the same in return.

Reply:
If the intention of DensityInterface is to not support this use case (which is OK but should be stated and communicated more clearly probably, in particular since discussions involved likelihoods), I think it will be unlikely that `logdensityof` is adopted in the near future in Turing.

Reply:
> I would argue that the PDF and the likelihood are very different things, semantically and mathematically, so we should not use logdensityof(d, x) to get the one or the other depending on the type of x.
> Ah, no, that's also not so nice, since the argument of a likelihood density are the parameters of the distribution, not the data.

I guess there might be another difference in our interpretation of `loglikelihood` here: in Turing and in my view, `loglikelihood(dist, x)` gives us just the log density of the conditional distribution `law(x | theta)` where `theta` are the parameters of the distribution `x`. Whereas it seems you think about the log-likelihood function `L(theta) = p(x | theta)` of the parameters `theta` for a fixed set of observations `x`. I agree that in the latter case the use of `logdensityof(dist, x)` seems weird but I would argue if it is viewed as a distribution of `x` it seems quite natural.

Reply:
I do think we should use a wrapper, but `Likelihood` is the wrong name: What we mean here (for Turing, etc.) is the Distribution/PDF over a set of possible outcomes. So it's like the PDF of an implicit product distribution, actually. So we could write

```julia
struct DatasetDensity{D<:Distribution}
    distribution::D
end

@inline DensityInterface.hasdensity(d::DatasetDensity) = true

@inline DensityInterface.logdensityof(d::DatasetDensity, X) = likelihood(d.distribution, X)
```

And Turing could just use `logdensity(DatasetDensity(d), X)`. This would be clean.

And `logdensityof(d::Distribution, x)` should actually call `logpdf(d, x)`, *not* `loglikelihood`.

I'm sorry, I think I contributed to the confusion: When @devmotion proposed changing to `loglikelihood` I misunderstood/misread things a bit, I think.

Reply:
> another difference in our interpretation of loglikelihood

Well, in the Bayesian inference world we typically think of the data as fixed, and when people say log-likelihood they do indeed mean `L(theta) = p(x | theta)`. My main use case for supporting `logdensityof` for Distributions is to get the log-density of a prior distribution, and the log-density of a likelihood, both as a function of theta.

But I get that there are other use cases, and of course we should strive to serve those as well.

However, I think we shouldn't muddle the semantics of `logdensityof(d, x)` too much. The definition is that it gives us the log-value of the density of `d` (if `d` is not a density itself) in respect to it's base measure at the point x. So for a distribution, that would be it's log-PDF at x.

> log density of the conditional distribution law(x | theta)

But this distribution is not the original distribution `d(theta)` - instead, it's a product distribution over `d` (`d` repeated n-times for n elements in `x`), correct?

For a single point I think it's clear that we want `logdensityof(d, x) == logpdf(d, x) == loglikelihood(d, x)` (in the sense of `Distributions.loglikelihood`), correct?

The question is, given multiple points, should we define the "natural" density of `d` to be the density of it's product distribution? It seems a bit too magical too me. `logdensityof(ProductDensity(d), X)` would be a way to state that we want that explicitly. Would there be any problems for Turing with this approach?

Reply:
Thanks @oschulz, I think this would be a clean solution! I was debating with myself back and forth whether one should use `loglikelihood` in the definition of `logdensityof`. I was not completey happy about it but a) it seemed a more general and, e.g. in Turing, more useful definition than `logpdf` and b) its implementation in Distributions satisfies properties that we expect from `logdensityof` but do not hold for `logpdf` for all inputs (e.g., always returns a real number). However, I agree that probably it is better to separate the "base case" and the implicit product distribution of iid samples more clearly.

I'm not completely sure if `DatasetDensity` is the best name. The name `DatsetDensity` seems to suggest that `X` is a set of multiple samples (or at least a set of samples) but it is fine to call `loglikelihood` with a single sample (it just falls back to `logpdf`), and this property is used in Turing as well. Maybe something like `IIDDensity` could be an alternative that does not imply necessarily that `X` are multiple samples?

Overall it seems we agree that we should implement the base case as
```julia
hasdensity(::Distribution) = true
logdensityof(d::Distribution, x) = logpdf(d, x)
densityof(d::Distribution, x) = pdf(d, x)
```
and the iid case with a dedicated density such as suggested above (preferably without the `@inline`, at least in `logdensityof` :stuck_out_tongue:). I think we should not be worried that we can't determine a priori which types of `x` (or `X`) are valid for the distribution at hand - the implementations (or rather non-existing implementations) of `logpdf` and `loglikelihood` for invalid sample(s) will take care of it.

Reply:
> I'm not completely sure if DatasetDensity is the best name

Me neither, was just the best I could come up with quickly for a prototype. :-) I like `IIDDensity`.

I'll implement it, can you just review JuliaMath/DensityInterface.jl#6 quickly? We need approximate comparison in `test_density_interface` for this, since `exp(logpdf(d, x))` is not always exactly `pdf(d, x)`.

Reply:
One problem with the name `IIDDensity`: We don't really require the sample to be independent, just to be identically distributed, right? But on the other hand, `IDDensity` sound like a density of IDs ...

Reply:
>  We don't really require the sample to be independent, just to be identically distributed, right?

We do require independence, don't we? Otherwise it is possible that the density of the samples can't be factorized and we don't end up with `loglikelihood(d, samples) = sum(logpdf(d, x) for x in samples)`.

Reply:
Ah - no, I mean the sample could, for example, result from an MCMC sampling. Even it it was a faithful sampling of `d`, the samples wouldn't be independent, since they would stem from a Markov chain.

Reply:
Ah OK, you're concerned with the fact that `X` for which we perform the evaluation is not actually a set of iid samples but of correlated samples? I would say this is irrelevant here, in the same way as it is e.g. irrelevant if a single sample `x` is actually a sample from the distribution for which we evaluate the density. IMO the name `IIDDensity` would still be correct since we perform the evaluation under the assumption that samples are iid.

Reply:
Ok, let's go with `IIDDensity` then, I don't have a better proposal anyway. :-)